base/abstractdict.jl

```diff
@@ -502,108 +502,149 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
+        d = IdDict{K,V}()
         for (k,v) in itr; d[k] = v; end
         d
     end
 
-    function ObjectIdDict(pairs::Pair...)
-        d = ObjectIdDict()
+    function IdDict{K,V}(pairs::Pair...) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
+        for (k,v) in pairs; d[k] = v; end
+        d
+    end
+
+    IdDict{K,V}(d::IdDict{K,V}) where {K, V} = new{K,V}(copy(d.ht))
+    function IdDict{K,V}(d::IdDict{K}) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
         for (k,v) in pairs; d[k] = v; end
         d
     end
+end
+const _IdDict = IdDict{Any,Any} # this is needed to make src/dump.c and src/staticdata.c work
+
+IdDict() = IdDict{Any,Any}()
+IdDict(kv::Tuple{}) = IdDict()
+
+IdDict(ps::Pair{K,V}...)           where {K,V} = IdDict{K,V}(ps)
+IdDict(ps::Pair{K}...)             where {K}   = IdDict{K,Any}(ps)
+IdDict(ps::(Pair{K,V} where K)...) where {V}   = IdDict{Any,V}(ps)
+IdDict(ps::Pair...)                            = IdDict{Any,Any}(ps)
 
-    ObjectIdDict(o::ObjectIdDict) = new(copy(o.ht))
+function IdDict(kv)
+    try
+        dict_with_eltype((K, V) -> IdDict{K, V}, kv, eltype(kv))
+    catch e
+        if !applicable(start, kv) || !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError(
+                "IdDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow(e)
+        end
+    end
 end
 
-empty(d::ObjectIdDict, ::Type{Any}, ::Type{Any}) = ObjectIdDict()
+empty(d::IdDict, ::Type{K}, ::Type{V}) where {K, V} = IdDict{K,V}()
 
-function rehash!(t::ObjectIdDict, newsz = length(t.ht))
-    t.ht = ccall(:jl_idtable_rehash, Any, (Any, Csize_t), t.ht, newsz)
-    t
+function rehash!(d::IdDict, newsz = length(d.ht))
+    d.ht = ccall(:jl_idtable_rehash, Any, (Any, Csize_t), d.ht, newsz)
+    d
 end
 
-function sizehint!(t::ObjectIdDict, newsz)
+function sizehint!(d::IdDict, newsz)
     newsz = _tablesz(newsz*2)  # *2 for keys and values in same array
-    oldsz = length(t.ht)
+    oldsz = length(d.ht)
     # grow at least 25%
     if newsz < (oldsz*5)>>2
-        return t
+        return d
     end
-    rehash!(t, newsz)
+    rehash!(d, newsz)
 end
 
-function setindex!(t::ObjectIdDict, @nospecialize(v), @nospecialize(k))
-    if t.ndel >= ((3*length(t.ht))>>2)
-        rehash!(t, max(length(t.ht)>>1, 32))
-        t.ndel = 0
+function setindex!(d::IdDict{K,V}, v, k::K) where {K, V}
+    v = convert(V, v)
+    if d.ndel >= ((3*length(d.ht))>>2)
+        rehash!(d, max(length(d.ht)>>1, 32))
+        d.ndel = 0
     end
-    t.ht = ccall(:jl_eqtable_put, Array{Any,1}, (Any, Any, Any), t.ht, k, v)
-    return t
+    d.ht = ccall(:jl_eqtable_put, Array{Any,1}, (Any, Any, Any), d.ht, k, v)
+    return d
 end
 
-get(t::ObjectIdDict, @nospecialize(key), @nospecialize(default)) =
-    ccall(:jl_eqtable_get, Any, (Any, Any, Any), t.ht, key, default)
+function get(d::IdDict{K,V}, key::K, default) where {K, V}
```

Comment:
The key doesn't need to be of type `K`, since it's ok for it not to be found here. The `@nospecialize` annotations can also probably be retained.

Reply:
Ok, so you'd prefer a KeyError to a MethodError.  Would it then also make sense to throw a KeyError for `setindex!`?

Also, should I retain all `@nospecialize` or just the one you mentioned here?

Reply:
Maybe this is similar enough to the `ArgumentError("$key0 is not a valid key for type $K")` that `Dict` throws that we should use the same error here?

And yes, we should keep all the `@nospecialize`. The goal of this should just be to allow type parameters in `ObjectIdDict` with the minimum changes.